11 add testing harness

exachixkitsune opened the pull request:
Here's hoping this all works properly...

This should be an automated test framework, with a workflow entity to make this go. I haven't actually tested it yet, but fingers crossed the workflow goes.
I am also expecting the test cases to fail after the merge happens due to code changes.

May well need the integration to occur before working out if it reports properly. I've never done this before. Time to see how it goes!

Closes #11 
Potentially fixes #27

Commit:
Fundaments of test framework

Implementation of required components for pytest and pytest-cov.
tools.test based on tools.lint to generate results in the same way.
Commit consists ONLY of the critical components for running tests, no actual tests provided.

Commit:
Initial experimental tests for demonstration purposes.

Commit to be overwritten in the future with proper quests.

exachixkitsune committed:
Refinement of automated test harness.

test.py now setup to extract information as appropriate.
Untested addition of workflow. copied from lint.yml.
Project coverage level reduced for initial testing.

exachixkitsune committed:
New test cases to bring coverage level up.

Test cases primarily focus on running the example yaml files through the loader.

exachixkitsune commented:
Also I've just realised the reporting for the workflow actions may want particular formats; can revise if necessary.

exachixkitsune committed:
Encompassing tests in lint, and resolving issues

Various common elements in tools and tests are now combined into "common".
  > common Annotation class
  > Printing annotation list into the window.
  > common running of utility
Resolution of linting errors:
  > name issue in test.py, also changed lint.py for parity.
  > Many changes in tests due to lint requirements
Notable overrides of flake8:
  > tools.test.parse_pytest_out has an increased complexity level
Notable overrides of pylint:
  > "too few public methods" and "no self use" suppressed in test files, as this is expected

javajawa commented on tests/common.py:
```diff
@@ -0,0 +1,43 @@
+from mewbot.io.http import HTTPServlet
+
+
+class BaseTestClassWithConfig(ABC):
+    # pylint: disable=R0903
+    # Too few public methods, because this is always used elsewhere
+    fname: str
+    config: ConfigBlock
+    doctype: ComponentKind
+    config_toload = True
+
```
I love the concept of this class! Very nice way to load a component from YAML.

This comment covers two additional thoughts:
 - I find overriding properties to be an unclear way to give info to a super class
 - We could do some fun things with the type system

For the first, I'd recommend setting up some stuff in `__init__`:
```suggestion
class BaseTestClassWithConfig(ABC):
    """Caches configs and components for testing"""
    fname: str
    doctype: ComponentKind
    _config: Optional[ConfigBlock] = None
    config_toload = True

    def __init__(self, filename: str, kind: ComponentKind) -> None:
        self.fname = filename
        self.doctype = kind

    @property
    def config(self) -> ConfigBlock:
        self.cache_config()
        return self._config

    @property
    def component(self) -> Component:
        self.cache_config()
        return load_component(self.config)
```

For `ExampleHttpPostIOConfig`, this would make the class look more like:
```python
class ExampleHttpPostIOConfig(BaseTestClassWithConfig):
    def __init__(self) -> None:
        super().__init__("examples/trivial_http_post.yaml", ComponentKind.IO_CONFIG)

    @property
    def component(self) -> HTTPServlet:
        component = super().component
        assert isinstance(component, HTTPServlet)
        return component
```

The typing idea in the second bullet is around using the Generics in python typing, which might look something like:

```suggestion
T = TypeVar("T", bound=Component, covariant=True)


class BaseTestClassWithConfig(ABC, Generic[T]):
    _config: ConfigBlock
    _component: Optional[T] = None

    def __init__(self, filename: str, implementation: Type[T]) -> None:
        with open(filename, "r", encoding="utf-8") as config:
            self._config = next([document for document in yaml.load_all(config, Loader=yaml.CSafeLoader) if document["implement"] == implementation.__name__])

    @property
    def config(self) -> ConfigBlock:
        return self._config

    @property
    def component(self) -> T:
        if not self._component:
            component = load_component(self._config)
            assert isinstance(component, T), "Config loaded incorrect type"
            self._component = component
        return self._component
```

Which should mean that doing something like `BaseTestClassWithConfig("examples/my_example.yaml", mewbot.discord.DiscordIO)` will type check as returns in a `DiscordIO` without any extra work

javajawa replied:
^ Note: I wrote all that from memory, bug me if things don't actually work

ajCameron replied:
I second the above on the form - very nice!

javajawa replied:
So it turns out that `pytest` will ignore classes with `__init__`, but I have been able to do "load this specific object from a config file" which looks like

```python
class TestLoaderHttpsPost(BaseTestClassWithConfig[HTTPServlet]):
    config_file: str = "examples/trivial_http_post.yaml"
    implementation: Type[HTTPServlet] = HTTPServlet

    # Test this working
    def test_working(self) -> None:
        assert self.config["implementation"] == "mewbot.io.http.HTTPServlet"

    # Test that the loading is accurate
    def test_loading_component_type(self) -> None:
        assert isinstance(self.component, HTTPServlet)
```

javajawa commented on tests/common.py:
```diff
@@ -0,0 +1,43 @@
+    config_toload = True
+
+    def cache_config(self) -> None:
+        if self.config_toload:
+            with open(self.fname, "r", encoding="utf-8") as config:
+                for document in yaml.load_all(config, Loader=yaml.CSafeLoader):
+                    if document["kind"] == self.doctype:
+                        self.config = document
+                        self.config_toload = False
```
I'd invert the condition here to reduce indentation, and also exit out once we match an item
 ```suggestion
        if not self.config_toload:
            return
        with open(self.fname, "r", encoding="utf-8") as config:
            for document in yaml.load_all(config, Loader=yaml.CSafeLoader):
                if document["kind"] == self.doctype:
                    self.config = document
                    self.config_toload = False
                    return
```

javajawa commented on tests/test_core.py:
```diff
@@ -0,0 +1,62 @@
+    IOConfigInterface,
+)
+
+# pragma pylint: disable=R0903
+#  Disable "too few public methods" for test cases - most test files will be classes used for
+#  grouping and then individual tests alongside these
+# pragma pylint: disable=R0201
+#  Disable "no self use" for functions. These functions will not be used internally as they are
+#  automatically called by pytest as it seeks and searches for tests.
```
You can avoid having to do this be removing the class, and just having a list of functions in the file. As far as I can tell, having the tests in a class doesn't do anything in terms of the information pytest returns.

exachixkitsune replied:
(Regarding fixing, it probably ends up as a convention thing; I personally would group tests into classes, particularly because that allows commonality with some configuration settings for some cases, and then the other cases match the same convention - but that's a Kitsune stylistic thing.)

The class containing the tests is written into output messages when the tests fail.

For example, if I simulate an error in the function `test_componentkind_interface_map_ioconfig`, it says: `TestComponent.test_componentkind_interface_map_ioconfig` and `FAILED tests/test_core.py::TestComponent::test_componentkind_interface_map_ioconfig - AssertionError:...`

rather than removing `test_componentkind_interface_map_ioconfig` from it's class, instead I get: `test_componentkind_interface_map_ioconfig` and `FAILED tests/test_core.py::test_componentkind_interface_map_ioconfig - AssertionError:...`

Which only really matters for visually seeing where errors are, rather than the automated output system in `tools/test.py`

ajCameron replied:
I find grouping tests into classes helpful in a number of ways
 - It allows for (easier) automated test generation. This can be particularly useful when trying to optimize (later - "premature optimization" and all) when, for example, you have a newly written optimized component which you want to test with a bunch of other configs to make sure everything all works
 - Given this program's going to be all config driven, I suspect being able to subclass tests for slightly different configs will be really helpful
 - It allows (again, possibly just me personally) to derive useful stuff from them more easily (e.g. "this started life as a test suite. Now it's part of my optimization and profiling toolchain")

So yes! I am opinionated in favour of test classes.

javajawa commented on tools/lint.py:
```diff
@@ -60,23 +36,11 @@ def run_tool(self, name: str, *args: str) -> subprocess.CompletedProcess[bytes]:
 
-        if run.returncode:
-            self.success = False
+        self.success = run_result["success"]
```
This is technically a change in behaviour -- previously, we could never go from false to true, but this permits this.
When fiddling locally, I used
```python
self.success = self.success and (run_result.returncode == 0)
```
(I also took out the typed dict with the `success` property and just returned the completed process directly).

That said, looking at this function, there is the possibilty of extracting an abstract class from thsese two, so that success logic isn't duplicated...

javajawa commented on tools/test.py:
```diff
@@ -0,0 +1,113 @@
+                    )
+
+            if ("FAIL Required test coverage" in output) and ("not reached." in output):
+                yield Annotation("error", "tools/test.py", 1, 1, "CoverageError", output)
```
The annotation probably wants to go against the file what was not covered?

exachixkitsune replied:
The threshold for code coverage appears to be overall coverage rather than specific file coverage; thus there is no specific file to relate to.

ajCameron replied:
Perhaps split the difference and annotate files that are _particularly_ bad? E.g. files with no test coverage at all. But I think that's a feature which could be left till later.

ajCameron commented on tools/common.py:
```diff
@@ -0,0 +1,68 @@
+        run.stdout = b""
+        run.stderr = b""
+
+    return {"success": success, "completedProcess": run}
```
Looks good to me!

ajCameron reviewed (commented):
Looks good to me!

ajCameron reviewed (dismissed):
If my comments seem terse, that's mostly because @javajawa said most of what I was thinking. All looks good from here.

javajawa committed:
Fix bahviour of tools folder

javajawa committed:
Update test code with new package names

javajawa committed:
Refactor how tools are run with better abstraction

javajawa committed:
Refactor tests to load config in a more consistent manner

Tests now specify the config file to load and the implementation of the object they are testing.

javajawa commented:
@exachixkitsune I took the liberty of implementing my suggestion for the typed loader, and did some housekeeping.

exachixkitsune commented:
> @exachixkitsune I took the liberty of implementing my suggestion for the typed loader, and did some housekeeping.

So I see; looks good to me at the moment.

javajawa committed:
Fix testing github workflow

javajawa committed:
Remove python 3.8 as a supported version

javajawa commented:
@exachixkitsune @ajCameron I think this is now "done" -- not sure I'm happy with what I've done with the history, but it does preserve the conversations in this thread pretty well.

ajCameron commented:
It looks good to me! I'll do a final scan over the code and then approve.

ajCameron reviewed (approved):
All looks good to me! Thanks for getting this done.
